alf/examples/ppo_icm_super_mario_intrinsic_only.gin

```diff
@@ -15,27 +15,38 @@ suite_mario.load.state="Level1-1"
 # algorithm config
 CONV_LAYER_PARAMS=[(32,8,4),(64,4,2),(64,3,1)]
 
-create_ac_algorithm.use_rnns=False
-create_ac_algorithm.learning_rate=3e-5
-create_ac_algorithm.use_icm=True
-create_ac_algorithm.encoding_conv_layers=%CONV_LAYER_PARAMS
-create_ac_algorithm.encoding_fc_layers=(256,)
+actor/ActorDistributionNetwork.input_tensor_spec=%observation_spec
+actor/ActorDistributionNetwork.output_tensor_spec=%action_spec
+actor/ActorDistributionNetwork.fc_layer_params=(256,)
+actor/ActorDistributionNetwork.activation_fn=@tf.nn.elu
+actor/ActorDistributionNetwork.conv_layer_params=%CONV_LAYER_PARAMS
+CategoricalProjectionNetwork.logits_init_output_factor=1e-10
+actor/ActorDistributionNetwork.discrete_projection_net=@CategoricalProjectionNetwork
 
-EncodingNetwork.activation_fn=@tf.nn.elu
-ICMAlgorithm.hidden_size=256
 
-# value network layers
-ValueNetwork.conv_layer_params=%CONV_LAYER_PARAMS
-create_ac_algorithm.value_fc_layers=(256,256)
-ValueNetwork.activation_fn=@tf.nn.elu
+value/ValueNetwork.input_tensor_spec=%observation_spec
+value/ValueNetwork.fc_layer_params=(256,256)
+value/ValueNetwork.conv_layer_params=%CONV_LAYER_PARAMS
+value/ValueNetwork.activation_fn=@tf.nn.elu
 
-# actor network layers
-ActorDistributionNetwork.conv_layer_params=%CONV_LAYER_PARAMS
-create_ac_algorithm.actor_fc_layers=(256,256)
-ActorDistributionNetwork.activation_fn=@tf.nn.elu
-CategoricalProjectionNetwork.logits_init_output_factor=1e-10
-ActorDistributionNetwork.discrete_projection_net=@CategoricalProjectionNetwork
+ac/Adam.learning_rate=3e-5
+
+icm/EncodingNetwork.input_tensor_spec=%observation_spec
+icm/EncodingNetwork.conv_layer_params=%CONV_LAYER_PARAMS
+icm/EncodingNetwork.fc_layer_params=(256,)
+icm/EncodingNetwork.activation_fn=@tf.nn.elu
+icm/to_tensor_spec.shape=(256,)
```

Comment:
should use a macro for this as it's should be same as  icm/EncodingNetwork.fc_layer_params.
Some other gin files for ICM may also have this problem

Reply:
feature_spec is (fc_layer_params[-1],) for a FC layers encoding network,  
They happen to be equal here as len(fc_layer_params)=1

but gin does not support config below: (Line3 , visit list element)
 
```
icm/encoding_fc_layers=(200,)
icm/EncodingNetwork.fc_layer_params=%icm/encoding_fc_layers
icm/to_tensor_spec.shape=(%icm/encoding_fc_layers[-1],)
ICMAlgorithm.feature_spec=@icm/to_tensor_spec()
```

Reply:
since basically function "to_tensor_spec" is only used for ICM, maybe we can remove function "to_tensor_spec", use 
`ICMAlgorithm.feature_shape=icm/encoding_fc_layers`, and add
`feature_spec=tf.TensorSpec(shape=feature_shape, dtype=tf.float32)` in icm_alg.py,

Reply:
algorithms constructor using `tf.TensorSpec`  to specifiy input or output spec, we should keep interface consistency

Reply:
I would be much better If gin can accept function calling with arguments

Reply:
Since we know it's same as the 200 used in icm/encoding_fc_layers=(200,),  we should use a macro so that the error of changing only one of them won't occur.

> feature_spec.shape is (fc_layer_params[-1],) for a FC layers encoding network,
> They happen to be equal here as len(fc_layer_params)=1
> 
> but gin does not support config below: (Line3 , visit list element)
> 
> ```
> icm/encoding_fc_layers=(200,)
> icm/EncodingNetwork.fc_layer_params=%icm/encoding_fc_layers
> icm/to_tensor_spec.shape=(%icm/encoding_fc_layers[-1],)
> ICMAlgorithm.feature_spec=@icm/to_tensor_spec()
> ```

Reply:
> I would be much better If gin can accept function calling with arguments

it would be more flexible and easier to use , i'll  see if it is easy to extend.